docs/design/namespaces.md

```diff
@@ -1,193 +1,219 @@
+to define *Hard* resource usage limits that a *Namespace* may consume.
+See [Admission control: Resource Quota](admission_control_resource_quota.md)
 
-The k8s API server verifies the *Namespace* on resource creation matches the *{ns}* on the path.
+### Finalizers
 
-The k8s API server will enable efficient mechanisms to filter model resources based on the *Namespace*.  This may require
-the creation of an index on *Namespace* that could support query by namespace with optional label selectors.
+Upon creation of a *Namespace*, the creator may provide a list of immutable *Finalizer* objects.
 
-The k8s API server will associate a resource with a *Namespace* if not populated by the end-user based on the *Namespace* context
-of the incoming request.  If the *Namespace* of the resource being created, or updated does not match the *Namespace* on the request,
-then the k8s API server will reject the request.
+```
+// NamespaceSpec describes the attributes on a Namespace
+type NamespaceSpec struct {
+  // Finalizers is a list of named finalize actions that must complete prior to namespace deletion.
+  Finalizers []FinalizerName
+}
 
-TODO: Update to discuss k8s api server proxy patterns
+type FinalizerName string
 
-## k8s storage
+// These are internal finalizers to Kubernetes, must be qualified name unless defined here
+const (
+  FinalizerKubernetes FinalizerName = "kubernetes"
+)
 
-A namespace provides a unique identifier space and therefore must be in the storage path of a resource.
+// NamespaceStatus is information about the current status of a Namespace.
+type NamespaceStatus struct {
+  // Phase is the current lifecycle phase of the namespace.
+  Phase NamespacePhase `json:"phase,omitempty"`
+  // Finalizers is the list of named finalize actions that have completed
+  Finalizers []FinalizerName `json:"finalizers"`
+}
+```
 
-In etcd, we want to continue to still support efficient WATCH across namespaces.
+A *FinalizerName* is a qualified name.
 
-Resources that persist content in etcd will have storage paths as follows:
+A *Finalizer* denotes a named entity that must release it's claim to the *Namespace* resource
+prior to its deletion.
 
-/registry/{resourceType}/{resource.Namespace}/{resource.Name} 
+The *Finalizer* pattern allows for asynchronous controllers that are watching cluster state
+to take part in distributed graceful termination of *Namespace* managed resources that are
+unknown to the cluster.
 
-This enables k8s service to WATCH /registry/{resourceType} for changes across namespace of a particular {resourceType}.
+During *Namespace* termination, each *Finalizer* must perform a *finalize* operation before
+the cluster is permitted to delete the *Namespace*.
 
-Upon scheduling a pod to a particular host, the pod's namespace must be in the key path as follows:
+A *finalize* operation updates the *NamespaceStatus.Finalizers* list noting that the named
+finalizer has completed its operation.
```

Comment:
I don't want to count on Status fields being persisted -- persistence is an availability and performance optimization. Status must be observable state. If you imagine the finalizer working like a scheduling looking for pods with no host assigned, then I could imagine this working -- it would be a robust approach. Alternatives would be for finalize to remove the finalizer from the list in the spec, or to update the spec with a completion timestamp or something.

You imagine `finalize` to be a subresource, similar to `binding`? That seems fine.